### PR TITLE
Music: switch music type in settings

### PR DESCRIPTION
**Pull request by idshibanov**
Switching between original castle tracks, expansion (from heroes2x.agg) and cd audio. CD audio not fully supported yet, however option works.

Also I matched Succession Wars tracks to the intended CD tracks (if you don't have CD here it is: https://www.youtube.com/watch?v=VRcprUCjwDc). As a widely known issue there's no actual Sorceress XMI file in the HEROES2.AGG, so it will fall back to either expansion theme or Warlock theme (as it did in OG).

It looks exactly like original:
image

### CODE REVIEW

**Comment by ihhub** [comment]
Hi @idshibanov , I also sent you an invitation. Please check your mailbox and do not hesitate to ask any questions if any :)

File: fheroes2.cfg [hunk 1]
```diff
@@ -13,8 +13,8 @@
 # sound: on off
 sound = on
 #
-# music: on,midi,cd,ext,off
-music = on
+# music: original, expansion, cd, ext
```
[the remark this answers is not in the document]
Comment (idshibanov) [reply fheroes2.cfg:16]: Sure, makes sense. External music code also might need an update in the future and a bit more documentation, but I doubt anyone uses this feature right now.

File: src/fheroes2/agg/agg.cpp [hunk 2]
```diff
@@ -1598,7 +1598,7 @@ void AGG::LoadWAV( int m82, std::vector<u8> & v )
 void AGG::LoadMID( int xmi, std::vector<u8> & v )
 {
     DEBUG( DBG_ENGINE, DBG_INFO, XMI::GetString( xmi ) );
-    const std::vector<u8> & body = ReadChunk( XMI::GetString( xmi ) );
+    const std::vector<u8> & body = ReadChunk( XMI::GetString( xmi ), xmi >= XMI::MIDI_ORIGINAL_KNIGHT );
```
[the remark this answers is not in the document]
Comment (idshibanov) [reply src/fheroes2/agg/agg.cpp:1601]: Yes, they are, even on CDs (Succession Wars vs Gold version). Also track numbers between MIDI and CD audio match.

HEROES2X.AGG only contains 6 music files: MIDI0005 to MIDI0010 (cities), the rest of them are pulled from base file. That's why I added them to the end of the enum to clearly separate from the rest and to be able to store both in AGG cache.

File: src/fheroes2/agg/xmi.cpp [hunk 3]
```diff
@@ -30,39 +30,60 @@ namespace XMI
...
+        {MIDI0013, "MIDI0013.XMI"},
+        {MIDI0014, "MIDI0014.XMI"},
+        {MIDI0015, "MIDI0015.XMI"},
+        {MIDI0017, "MIDI0017.XMI"},
+        {MIDI0018, "MIDI0018.XMI"},
+        {MIDI0042, "MIDI0042.XMI"},
+        {MIDI0043, "MIDI0043.XMI"},
+        {MIDI_ORIGINAL_KNIGHT, "MIDI0009.XMI"},
+        {MIDI_ORIGINAL_BARBARIAN, "MIDI0007.XMI"},
+        {MIDI_ORIGINAL_SORCERESS, "MIDI0005.XMI"}, // for consistency
```
[the remark this answers is not in the document]
Comment (idshibanov) [reply src/fheroes2/agg/xmi.cpp:53]: Sure, I'll add the explanation here. XMI files mapped here are already corrected to match the CD.

File: src/fheroes2/system/settings.cpp [hunk 4]
```diff
@@ -1263,6 +1286,12 @@ void Settings::SetMusicVolume( int v )
     music_volume = 10 <= v ? 10 : v;
 }
 
+/* music volume: 0 - 10 */
+void Settings::SetMusicType( int v )
+{
+    _musicType = MUSIC_CDROM <= v ? MUSIC_CDROM : static_cast<MusicSource>( v );
```
[the remark this answers is not in the document]
Comment (idshibanov) [reply src/fheroes2/system/settings.cpp:1292]: Game handles it fine. Right now there will be no music when you switch to CDROM. However, even on SDL1 with WITH_AUDIOCD compile flag it doesn't work too well. Fixes for it require its own PR.

**Comment by idshibanov** [comment]
> Hi @idshibanov , I also sent you an invitation. Please check your mailbox and do not hesitate to ask any questions if any :)

Got it, thank you! I can start labeling things now :)

**Comment by idshibanov** [comment]
I think now this is ready to go. I might add a wiki page regarding music later on, so we can point people to it.

File: fheroes2.cfg [hunk 5]
```diff
@@ -13,8 +13,8 @@
 # sound: on off
 sound = on
 #
-# music: on,midi,cd,ext,off
-music = on
+# music: original, expansion, cd, external
+music = original
```
Comment (ihhub) [inline fheroes2.cfg:17]: We should put **expansion** by default. It is a little awkward that we play songs from the original resources when we have expansion pack files.

File: src/fheroes2/dialog/dialog_system.cpp [hunk 6]
```diff
@@ -205,11 +215,23 @@ void Dialog::DrawSystemInfo( const Rects & rects )
...
+    text.Blit( rect3.x + ( rect3.w - text.w() ) / 2, rect3.y - text.h() - textOffset );
+
+    if ( conf.MusicType() == MUSIC_MIDI_ORIGINAL ) {
+        str = _( "MIDI" );
+    }
+    else if ( conf.MusicType() == MUSIC_MIDI_EXPANSION ) {
+        str = _( "MIDI Expansion" );
+    }
+    else if ( conf.MusicType() == MUSIC_CDROM ) {
+        str = _( "CD Stereo" );
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog_system.cpp:233]: We should allow this option if CD data is not present and for SDL 2 as we do not have a support for this if I am not wrong.

File: src/fheroes2/dialog/dialog_system.cpp [hunk 7]
```diff
@@ -205,11 +215,23 @@ void Dialog::DrawSystemInfo( const Rects & rects )
...
+    sprite3.Blit( rect3, display );
+    str = _( "Music Type" );
+    text.Set( str, Font::SMALL );
+    text.Blit( rect3.x + ( rect3.w - text.w() ) / 2, rect3.y - text.h() - textOffset );
+
+    if ( conf.MusicType() == MUSIC_MIDI_ORIGINAL ) {
+        str = _( "MIDI" );
+    }
+    else if ( conf.MusicType() == MUSIC_MIDI_EXPANSION ) {
+        str = _( "MIDI Expansion" );
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog_system.cpp:230]: We should not allow this option if we do not have expansion pack files.

File: src/fheroes2/system/settings.cpp [hunk 8]
```diff
@@ -602,19 +602,27 @@ bool Settings::Read( const std::string & filename )
...
-        if ( sval == "midi" ) {
+        if ( sval == "original" ) {
+            opt_global.ResetModes( GLOBAL_MUSIC );
+            opt_global.SetModes( GLOBAL_MUSIC_CD );
+            _musicType = MUSIC_MIDI_ORIGINAL;
+        }
+        else if ( sval == "expansion" ) {
             opt_global.ResetModes( GLOBAL_MUSIC );
             opt_global.SetModes( GLOBAL_MUSIC_MIDI );
+            _musicType = MUSIC_MIDI_EXPANSION;
```
Comment (ihhub) [inline src/fheroes2/system/settings.cpp:618]: Here we need to switch back to **original** if we do not have expansion files.

File: src/fheroes2/dialog/dialog_system.cpp (hunk 7 above)
Comment (idshibanov) [reply src/fheroes2/dialog/dialog_system.cpp:230]: It does file check for it right now when you switching between audio option in-game.

This part is display only, I think it should be as is so it's transparent and clear what game 'thinks'.

File: src/fheroes2/dialog/dialog_system.cpp (hunk 6 above)
Comment (idshibanov) [reply src/fheroes2/dialog/dialog_system.cpp:233]: This exact part is display only, so it should stay as is.

I'll update the music type switch to validate CD. However in the future (with full CD support) I thought to make this option visible at all times to suggest users to look for a game CD as it enhances the gameplay dramatically :)

Comment (ihhub) [reply src/fheroes2/dialog/dialog_system.cpp:236]: Yes, it is a good suggestion. However, it is not obvious for an user to think this way. We might need to change a label showing that there is no CD in the game. Then the user would know about this possibility.

**Comment by idshibanov** [comment]
> then we modify `void Settings::SetPriceLoyaltyVersion( void )` function to have an input **boolean** parameter. It it is true then we call `opt_global.SetModes( GLOBAL_PRICELOYALTY );` again (just in case), otherwise we need to call a new function let's say `RemovePriceLoyaltySettings()` where we check music setting.
> 
> What do you think?

Could work, I'll try this approach.

File: src/fheroes2/system/settings.cpp [hunk 9]
```diff
@@ -623,19 +625,28 @@ bool Settings::Read( const std::string & filename )
...
-    // music
+    // music source
+    _musicType = MUSIC_MIDI_ORIGINAL;
     sval = config.StrParams( "music" );
 
     if ( !sval.empty() ) {
-        if ( sval == "midi" ) {
+        if ( sval == "original" ) {
+            opt_global.ResetModes( GLOBAL_MUSIC );
+            opt_global.SetModes( GLOBAL_MUSIC_CD );
```
Comment (ihhub) [inline src/fheroes2/system/settings.cpp:635]: Should this value be `GLOBAL_MUSIC_MIDI` instead?

Comment (ihhub) [reply src/fheroes2/system/settings.cpp]: Hi @idshibanov , I made this change. Could you please confirm that it is a valid change?

Comment (idshibanov) [reply src/fheroes2/system/settings.cpp]: Yes, looks good now

**Comment by ihhub** [comment]
@idshibanov , thanks a lot for making these changes!

**Pull request merged**